pkg/image/apis/image/validation/whitelist/whitelister_test.go

```diff
@@ -0,0 +1,323 @@
+				{Registry: "localhost:5000", Namespace: "my", Name: "app"}:     nil,
+			},
+		},
+
+		{
+			name:      "allow any host:port with any transport",
+			transport: WhitelistTransportAny,
+			whitelist: mkAllowed(true, "*:*"),
+			hostnames: map[string]error{
+				"localhost:5000": nil,
+				"docker.io":      nil,
+				"localhost:443":  nil,
+				"localhost:80":   nil,
+				"example.com":    nil,
+			},
+			difs: map[imageapi.DockerImageReference]error{
+				{Registry: "docker.io", Namespace: "library", Name: "busybox"}: nil,
+				{Registry: "example.com", Namespace: "a", Name: "b"}:           nil,
+				{Registry: "localhost:80", Namespace: "ns", Name: "repo"}:      nil,
+				{Registry: "localhost:443", Namespace: "ns", Name: "repo"}:     nil,
+				{Registry: "docker.io", Name: "busybox"}:                       nil,
+				{Registry: "localhost:5000", Namespace: "my", Name: "app"}:     nil,
+			},
+		},
+
+		{
+			name:      "allow whitelisted with secure transport",
+			transport: WhitelistTransportSecure,
+			whitelist: mkAllowed(false, "localhost:5000", "docker.io", "example.com:*", "registry.com:80"),
+			hostnames: map[string]error{
+				"example.com:5000":         nil,
+				"example.com:80":           nil,
+				"example.com":              nil,
+				"localhost:443":            fmt.Errorf(`registry "localhost:443" not allowed by whitelist { "localhost:5000", "docker.io:443", "example.com:*", "registry.com:80" }`),
+				"localhost:5000":           nil,
+				"registry-1.docker.io:443": nil,
+				"registry.com:443":         fmt.Errorf(`registry "registry.com:443" not allowed by whitelist { "localhost:5000", "docker.io:443", "example.com:*", "registry.com:80" }`),
+				"registry.com:80":          nil,
+				"registry.com":             fmt.Errorf(`registry "registry.com" not allowed by whitelist { "localhost:5000", "docker.io:443", "example.com:*", "registry.com:80" }`),
+			},
+			difs: map[imageapi.DockerImageReference]error{
+				{Registry: "docker.io"}:            nil,
+				{Registry: "index.docker.io"}:      nil,
+				{Registry: "example.com"}:          nil,
+				{Registry: "docker.io"}:            nil,
+				{Registry: "localhost:5000"}:       nil,
+				{Registry: "registry.example.com"}: fmt.Errorf(`registry "registry.example.com" not allowed by whitelist { "localhost:5000", "docker.io:443", "example.com:*", "registry.com:80" }`),
+				{Name: "busybox"}:                  nil,
+			},
+		},
+
+		{
+			name:      "allow whitelisted with insecure transport",
+			transport: WhitelistTransportInsecure,
+			whitelist: mkAllowed(true, "localhost:5000", "docker.io", "example.com:*", "registry.com:80", "*.foo.com", "*domain.ltd"),
+			hostnames: map[string]error{
+				"a.b.c.d.foo.com:80": nil,
+				"domain.ltd":         nil,
+				"example.com":        nil,
+				"foo.com":            fmt.Errorf(`registry "foo.com" not allowed by whitelist { "localhost:5000", "docker.io:80", "example.com:*", "registry.com:80", and 2 more ... }`),
```

Comment:
on what basis is foo.com denied?  the whitelist has *.foo.com and allow insecure.

Reply:
The whitelisted registry has `.` as a prefix. Therefore it matches any sub-domain of `foo.com` but not the domain itself. I'd like to avoid additional logic on special treating `*.`. If the admin wants to match the domain as well, he can add the second entry without the leading `*.`.

Reply:
i'm still unclear on this one.  I would expect *.foo.com to match foo.com.

Reply:
discussed on irc